mempool/src/pool.rs

```diff
@@ -0,0 +1,903 @@
+use common::chain::transaction::TxInput;
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+// TODO Add a newtype macro
+
+// TODO this willbe defined elsewhere
+const MAX_BLOCK_SIZE_BYTES: usize = 1_000_000;
+
+pub const MEMPOOL_MAX_TXS: usize = 1_000_000;
+
+#[derive(PartialEq, Eq, Clone)]
+struct TxMempoolEntry {
+    tx: Transaction,
+    parents: BTreeSet<Rc<TxMempoolEntry>>,
+    is_replaceable: bool,
+}
+
+// TODO consider mocking the fees more accurately by referencing here a global mock Chain State
+// object
+fn compare_fees(this: &TxMempoolEntry, other: &TxMempoolEntry) -> Ordering {
+    // We want the comparison to be such that a collection is sorted if its elements are ordered in
+    // descending order
+    other.get_fee().cmp(&this.get_fee())
+}
+
+impl TxMempoolEntry {
+    fn new(tx: Transaction, parents: BTreeSet<Rc<TxMempoolEntry>>, is_replaceable: bool) -> Self {
+        Self {
+            tx,
+            parents,
+            is_replaceable,
+        }
+    }
+    //TODO this should really be sum of inputs minus sum of outputs
+    //But we don't yet have a way of summing of the inputs
+    //TODO handle the error bettter
+    fn get_fee(&self) -> Amount {
+        self.tx
+            .get_outputs()
+            .iter()
+            .map(|output| output.get_value())
+            .sum::<Option<_>>()
+            .expect("fee overflow")
+    }
+}
+
+impl PartialOrd for TxMempoolEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(compare_fees(self, other))
+    }
+}
+
+impl Ord for TxMempoolEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        compare_fees(self, other)
+    }
+}
+
+pub(crate) struct MempoolImpl<C: ChainStateView> {
+    store: MempoolStore,
+    chain_state: C,
+}
+
+struct MempoolStore {
+    // TODO some cloning can be avoided by having spender_txs be a map between references
+    txs_by_id: HashMap<H256, Rc<TxMempoolEntry>>,
+    txs_by_fee: BTreeSet<Rc<TxMempoolEntry>>,
+    spender_txs: HashMap<OutPoint, Rc<TxMempoolEntry>>,
+}
+
+impl MempoolStore {
+    fn new() -> Self {
+        Self {
+            txs_by_fee: BTreeSet::new(),
+            txs_by_id: HashMap::new(),
+            spender_txs: HashMap::new(),
+        }
+    }
+
+    // Checks whether the outpoint belongs to one of the txs in the mempool
+    fn contains_outpoint(&self, outpoint: &OutPoint) -> bool {
+        matches!(self.txs_by_id.get(&outpoint.get_tx_id().get()),
+            Some(entry) if entry.tx.get_inputs().len() > outpoint.get_output_index() as usize)
+    }
+
+    fn create_entry(&self, tx: Transaction) -> TxMempoolEntry {
+        let parents = tx
+            .get_inputs()
+            .iter()
+            .filter_map(|input| self.txs_by_id.get(&input.get_outpoint().get_tx_id().get()))
+            .cloned()
+            .collect::<BTreeSet<_>>();
```

Comment:
Good point. There is a check that all inputs/outpoints are available `verify_inputs_available` which is called in `validate_transaction`, and the `filter_map` is just there to avoid error handling (not the best practice now that I think of it). Let me know if `verify_inputs_available` addresses your concern in terms of correctness.

In terms of clarity: Given that the code as it stands raised your question, I think it is not clear enough. I will think of a way to make it clearer.

Reply:
The function responsible for resolving inputs is `verify_inputs_available`. It checks that every outpoint referenced by an input of the incoming transaction is either in the mempool, or in the `chain_state`, which to the best of my understanding should reflect the utxo set of confirmed transactions.

Reply:
Got rid of `create_entry` and changed the interface for `TxMempoolEntry::new` to avoid misuse of the latter.